### PR TITLE
PRM-5: Changeset validate_required_without functionality

### PR DESCRIPTION
**Pull request by PavelVesnin**
PRM-5: Changeset validate_required_without functionality

### CODE REVIEW

**Comment by PavelVesnin** [comment]
@gmile @AndrewDryga I need your opinion about `validate_required_without\4` function.
I think it could be simplified a little bit. Do you have some ideas?

File: lib/prm/doctor_api/api.ex [hunk 1]
```diff
@@ -34,6 +35,11 @@ defmodule PRM.DoctorAPI do
     doctor_changeset(doctor, %{})
   end
 
+  def search_doctor(attrs) do
+    attrs
+    |> DoctorSearch.validate()
```
Comment (PavelVesnin) [inline lib/prm/doctor_api/api.ex]: later I'll add Repo.find() here

**Comment by AndrewDryga** [comment]
This looks to complex, you receive a Map, call Enum.take to extract only fields that can be searched for, then apply a changeset that they contain valid data.

**Comment by gmile** [comment]
If I understand correctly what we're trying to accomplish here is:

1. if ANY field among required is specified – perform search,
2. otherwise – return error / empty result set.

Is that correct?

If yes, can we not add a validation as simple as the below? (stolen from [here](http://stackoverflow.com/questions/42212425/conditional-validation-in-ecto-for-or-1-of-2-fields-is-required)):

```elixir
def validate_any_is_present(changeset, fields) do
  if Enum.any?(fields, &present?(changeset, &1)) do
    changeset
  else
    add_error(changeset, hd(fields), "One of these fields must be present: #{inspect fields}")
  end
end

def present?(changeset, field) do
  value = get_field(changeset, field)
  value && value != ""
end
```

**Comment by PavelVesnin** [comment]
@gmile this is not what I need but I can use Enum.any? for simplifying code.
`validation_without` validates that field must be present and not empty only when any of the other specified fields are not present.

**Comment by PavelVesnin** [comment]
@gmile I can use it, thanks!

File: lib/prm/doctor_api/doctor_search.ex [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    %PRM.DoctorSearch{}
+    |> cast(attrs, fields)
+    |> validate_any_is_present(fields)
+  end
+
+  def validate_any_is_present(changeset, fields) do
+    if Enum.any?(fields, &present?(changeset, &1)) do
+      changeset
+    else
+      add_error(changeset, hd(fields), "One of these fields must be present: #{inspect fields}")
```
Comment (AndrewDryga) [inline lib/prm/doctor_api/doctor_search.ex]: I'd rather generate an error for each field that would say `:required, "This field is optional when any other valid search condition is set"`.

File: lib/prm/doctor_api/doctor_search.ex [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    end
+  end
+
+  defp present?(changeset, field) do
+    value = get_field(changeset, field)
```
Comment (AndrewDryga) [inline lib/prm/doctor_api/doctor_search.ex]: Maybe `get_change/1`? You don't need to look in data, it will always be empty.

Comment (AndrewDryga) [inline lib/prm/doctor_api/doctor_search.ex]: Or even split this function into two: `present?` and `empty?`

Comment (PavelVesnin) [inline lib/prm/doctor_api/doctor_search.ex]: Ok, let's check just changes

File: lib/prm/doctor_api/doctor_search.ex (hunk 2 above)
Comment (PavelVesnin) [inline lib/prm/doctor_api/doctor_search.ex]: In previous request it was. 
I'll revert validation message for each required_any field, but I think it's better to tell that
`:required, "This field is required only when any of %fields% fields are not present"`

**Comment by gmile** [comment]
Removed lib/prm/doctor_api/doctor_search.ex from the PR as it's not used

**Comment by gmile** [comment]
I have cleaned up the PR and would like to merge it asap. The coveralls API continues to return 500 error in travis builds despite my efforts. I'm giving up.

**Comment by PavelVesnin** [comment]
Coveralls failed three times on push because of 500, but pull request is ok.
Maybe let's merge it? @gmile

**Comment by PavelVesnin** [comment]
Fixed! Hooray!

**Pull request merged**